File: backend/app/routers/fundamentals.py
"""GET /api/fundamentals — the full fundamental set, with history.

Serves what db/load_fundamentals.py stores: the wide field list from the owner's Bloomberg pull,
Cary's Piotroski with its nine signals, and every annual observation on record.

TWO ROW KINDS, MERGED FOR DISPLAY BUT NEVER IN STORAGE
    An `annual` row carries statement figures dated by the filing's acceptance; a `snapshot` row
    carries market figures as of a fetch. They are stored apart because merging them invents a
    record that looks point-in-time and is not (migration 003 / db/load_fundamentals.py).

    A page, though, wants one line per company. So they are joined HERE, at the edge, with
    `as_of` fields naming when each half was true. The merge is a presentation decision and it is
    reversible; doing it in the database would not be.

HISTORY IS THE ANNUAL SERIES
    One row per filed period, oldest to newest, each with the acceptance date that makes it
    point-in-time safe. That is what makes "has this business been improving" answerable rather
    than a single snapshot repeated.
"""

from __future__ import annotations

import logging
from typing import Annotated, Any

from fastapi import APIRouter, HTTPException, Query

from app.db import DbUnavailable, connection
from app.validation import normalize_ticker

logger = logging.getLogger("agentic.api.fundamentals")

router = APIRouter(prefix="/api", tags=["fundamentals"])

# Ordered so the response reads like the owner's sheet rather than like the table's column order.
_ANNUAL_FIELDS = [
    "period_end", "known_at", "revenue_ttm", "ebitda_ttm", "eps_current", "eps_growth_yoy",
    "free_cash_flow", "fcf_yield", "capital_expenditure", "net_debt", "shares_outstanding",
    "gross_margin", "operating_margin", "net_margin", "ebitda_margin",
    "roe", "roc", "current_ratio", "quick_ratio", "debt_to_equity", "equity_to_assets",
    "ebitda_interest", "cash_conversion_cycle", "revenue_growth_yoy", "rd_to_revenue",
    "tangible_book_value_per_share", "piotroski_f_score", "piotroski_variant", "piotroski_signals",
    "derived_fields",
]
_MARKET_FIELDS = [
    "period_end", "known_at", "price", "market_cap", "pe_trailing", "pe_forward", "peg_ratio",
    "price_to_book", "price_to_sales", "price_to_tangible_book", "ev_to_ebitda", "dividend_yield",
    "beta", "week_52_high", "week_52_low", "avg_volume_30d", "analyst_target_price",
    "analyst_recommendation",
]


def _row_to_dict(cols: list[str], row) -> dict[str, Any]:
    out: dict[str, Any] = {}
    for name, value in zip(cols, row, strict=True):
        if value is None:
            out[name] = None
        elif name in ("piotroski_signals", "derived_fields"):
            out[name] = value  # already jsonb -> dict
        elif name in ("period_end", "known_at"):
            out[name] = str(value)
        elif name in ("piotroski_f_score", "avg_volume_30d"):
            out[name] = int(value)
        elif name in ("piotroski_variant", "analyst_recommendation"):
            out[name] = value
        else:
            out[name] = float(value)
    return out


def _fetch(conn, symbol: str) -> dict[str, Any] | None:
    sec = conn.execute(
        "SELECT id, symbol, name, sector, industry FROM securities WHERE upper(symbol)=upper(%s)",
        (symbol,),
    ).fetchone()
    if sec is None:
        return None
    sec_id, sym, name, sector, industry = sec

    annual_cols = ", ".join(_ANNUAL_FIELDS)
    history = [
        _row_to_dict(_ANNUAL_FIELDS, r)
        for r in conn.execute(
            f"SELECT {annual_cols} FROM fundamentals_snapshots"
            " WHERE security_id=%s AND period_type='annual' ORDER BY period_end DESC",
            (sec_id,),
        ).fetchall()
    ]
    market_cols = ", ".join(_MARKET_FIELDS)
    market_row = conn.execute(
        f"SELECT {market_cols} FROM fundamentals_snapshots"
        " WHERE security_id=%s AND period_type='snapshot' ORDER BY known_at DESC LIMIT 1",
        (sec_id,),
    ).fetchone()

    return {
        "symbol": sym,
        "name": name,
        "sector": sector,
        "industry": industry,
        # Named `market` and `latest_annual` rather than flattened into one object: a single blob
        # would hide that the two halves were true at different moments.
        "market": _row_to_dict(_MARKET_FIELDS, market_row) if market_row else None,
        "latest_annual": history[0] if history else None,
        "history": history,
        "periods_on_record": len(history),
    }


@router.get("/fundamentals")
def fundamentals_list(
    # Annotated rather than `symbols: str = Query("")`: with the latter the DEFAULT VALUE is a Query
    # object, so the parameter only becomes a string when FastAPI resolves it. Any internal caller
    # (a test, a script, another router) gets the sentinel and fails on .split(). The default here
    # is a real string.
    symbols: Annotated[str, Query(description="comma-separated; default: held names")] = "",
) -> dict:
    """One line per company, for the table. Defaults to what the account actually holds."""
    wanted = [s.strip().upper() for s in symbols.split(",") if s.strip()]
    try:
        with connection() as conn:
            if not wanted:
                wanted = [
                    r[0] for r in conn.execute(
                        "SELECT DISTINCT s.symbol FROM paper_portfolio_positions p"
                        " JOIN securities s ON s.id=p.security_id"
                        " JOIN paper_portfolios pp ON pp.id=p.portfolio_id"
                        " WHERE pp.kind='real' AND p.exit_date IS NULL ORDER BY 1"
                    ).fetchall()
                ]
            rows = [r for sym in wanted if (r := _fetch(conn, sym)) is not None]
            unknown = [s for s in wanted if s not in {r["symbol"].upper() for r in rows}]
    except DbUnavailable as exc:
        raise HTTPException(status_code=503, detail=f"The database is unavailable: {exc}") from None

    return {
        "meta": {
            "requested": wanted,
            # Named rather than silently dropped: a symbol missing from the table because it is not
            # in `securities` looks identical to one with no fundamentals until someone says so.
            "unknown_symbols": unknown,
            "count": len(rows),
            "piotroski_variant": "cary",
        },
        "rows": rows,
    }


@router.get("/fundamentals/{symbol}")
def fundamentals_one(symbol: str) -> dict:
    ticker = normalize_ticker(symbol)
    if ticker is None:
        raise HTTPException(status_code=400, detail="Not a valid ticker symbol.")
    try:
        with connection() as conn:
            row = _fetch(conn, ticker)
    except DbUnavailable as exc:
        raise HTTPException(status_code=503, detail=f"The database is unavailable: {exc}") from None
    if row is None:
        raise HTTPException(status_code=404, detail=f"{ticker} is not in the securities table.")
    if not row["history"] and not row["market"]:
        # A known security with nothing ingested is a DIFFERENT state from an unknown ticker, and
        # the page should say "not fetched yet" rather than "no such company".
        row["meta_note"] = (
            f"{ticker} is a known security but no fundamentals have been ingested for it yet. "
            f"Run db/load_fundamentals.py load --symbols {ticker}."
        )
    return row

File: frontend/src/lib/fundamentals.ts
// Types for `GET /api/fundamentals` — the full fundamental set with history.
//
// UNITS ARE NOT UNIFORM, AND THAT IS THE MAIN HAZARD HERE
//   Margins, growth rates, ROE/ROC, R&D intensity and equity/assets arrive as FRACTIONS (0.469 =
//   46.9%). Current ratio, quick ratio, debt/equity, interest coverage and per-share figures are
//   plain numbers. Cash conversion cycle is DAYS. Rendering a fraction with the percent formatter
//   for one field and not another is how a page ends up claiming 0.47% gross margin for Apple, so
//   every field below is routed through an explicit format in FIELD_GROUPS rather than guessed
//   from its value.
//
// TWO ROW KINDS
//   `market` is as of a fetch; `latest_annual` and `history` are as of a filing. The endpoint keeps
//   them apart and so does the page — each block states when it was true.

export interface PiotroskiSignals {
  /** Signals that came out true. Out of `evaluated`, NOT out of `of`. */
  score: number;
  /** The denominator the variant defines — 9 for Cary's. */
  of: number;
  /** How many of the nine could be computed at all. Below `of` means inputs were missing. */
  evaluated: number;
  /** True only when every signal was computable. A partial score is not an F-score. */
  complete: boolean;
  variant: string;
  periods: { prior: string | null; current: string | null };
  /** null = the inputs for this signal were missing, which is NOT the same as a failed signal. */
  signals: Record<string, boolean | null>;
}

export interface AnnualPeriod {
  period_end: string;
  /** The filing's acceptance date — what makes this row point-in-time safe. */
  known_at: string;
  revenue_ttm: number | null;
  ebitda_ttm: number | null;
  eps_current: number | null;
  eps_growth_yoy: number | null;
  free_cash_flow: number | null;
  fcf_yield: number | null;
  capital_expenditure: number | null;
  net_debt: number | null;
  shares_outstanding: number | null;
  gross_margin: number | null;
  operating_margin: number | null;
  net_margin: number | null;
  ebitda_margin: number | null;
  roe: number | null;
  roc: number | null;
  current_ratio: number | null;
  quick_ratio: number | null;
  debt_to_equity: number | null;
  equity_to_assets: number | null;
  ebitda_interest: number | null;
  cash_conversion_cycle: number | null;
  revenue_growth_yoy: number | null;
  rd_to_revenue: number | null;
  tangible_book_value_per_share: number | null;
  /** null when the nine signals could not all be computed — see PiotroskiSignals.complete. */
  piotroski_f_score: number | null;
  piotroski_variant: string | null;
  piotroski_signals: PiotroskiSignals | null;
  /** field -> the formula used, for anything we calculated rather than received. */
  derived_fields: Record<string, string> | null;
}

export interface MarketBlock {
  period_end: string;
  known_at: string;
  price: number | null;
  market_cap: number | null;
  pe_trailing: number | null;
  pe_forward: number | null;
  peg_ratio: number | null;
  price_to_book: number | null;
  price_to_sales: number | null;
  price_to_tangible_book: number | null;
  ev_to_ebitda: number | null;
  dividend_yield: number | null;
  beta: number | null;
  week_52_high: number | null;
  week_52_low: number | null;
  avg_volume_30d: number | null;
  analyst_target_price: number | null;
  analyst_recommendation: string | null;
}

export interface FundamentalsRow {
  symbol: string;
  name: string | null;
  sector: string | null;
  industry: string | null;
  market: MarketBlock | null;
  latest_annual: AnnualPeriod | null;
  /** Newest first. */
  history: AnnualPeriod[];
  periods_on_record: number;
  /** Present when the security is known but nothing has been ingested for it. */
  meta_note?: string;
}

export interface FundamentalsResponse {
  meta: {
    requested: string[];
    /** Requested symbols absent from `securities` — named rather than silently dropped. */
    unknown_symbols: string[];
    count: number;
    piotroski_variant: string;
  };
  rows: FundamentalsRow[];
}

export type Unit = "usd" | "usdCompact" | "fraction" | "ratio" | "days" | "shares" | "int";

export interface FieldSpec {
  key: keyof AnnualPeriod | keyof MarketBlock;
  label: string;
  unit: Unit;
  /** Higher is better (true), lower is better (false), or neither (undefined). Drives the trend
   *  arrow colour in the history table — and is left undefined wherever "better" is genuinely
   *  contestable (net debt in a rate-cutting cycle, capex for a company in a build-out). */
  higherIsBetter?: boolean;
}

/** The owner's sheet, grouped. Order within a group is the reading order, not the column order. */
export const ANNUAL_GROUPS: { title: string; fields: FieldSpec[] }[] = [
  {
    title: "Scale & growth",
    fields: [
      { key: "revenue_ttm", label: "Revenue", unit: "usdCompact", higherIsBetter: true },
      { key: "revenue_growth_yoy", label: "Revenue growth YoY", unit: "fraction", higherIsBetter: true },
      { key: "ebitda_ttm", label: "EBITDA", unit: "usdCompact", higherIsBetter: true },
      { key: "eps_current", label: "EPS", unit: "usd", higherIsBetter: true },
      { key: "eps_growth_yoy", label: "EPS growth YoY", unit: "fraction", higherIsBetter: true },
      { key: "shares_outstanding", label: "Shares out", unit: "shares", higherIsBetter: false },
    ],
  },
  {
    title: "Margins & returns",
    fields: [
      { key: "gross_margin", label: "Gross margin", unit: "fraction", higherIsBetter: true },
      { key: "operating_margin", label: "Operating margin", unit: "fraction", higherIsBetter: true },
      { key: "ebitda_margin", label: "EBITDA margin", unit: "fraction", higherIsBetter: true },
      { key: "net_margin", label: "Net margin", unit: "fraction", higherIsBetter: true },
      { key: "roe", label: "Return on equity", unit: "fraction", higherIsBetter: true },
      { key: "roc", label: "Return on capital", unit: "fraction", higherIsBetter: true },
    ],
  },
  {
    title: "Cash flow",
    fields: [
      { key: "free_cash_flow", label: "Free cash flow", unit: "usdCompact", higherIsBetter: true },
      { key: "capital_expenditure", label: "Capex", unit: "usdCompact" },
      { key: "fcf_yield", label: "FCF yield", unit: "fraction", higherIsBetter: true },
      { key: "cash_conversion_cycle", label: "Cash conversion cycle", unit: "days", higherIsBetter: false },
      { key: "rd_to_revenue", label: "R&D / revenue", unit: "fraction" },
    ],
  },
  {
    title: "Balance sheet",
    fields: [
      { key: "net_debt", label: "Net debt", unit: "usdCompact" },
      { key: "debt_to_equity", label: "Debt / equity", unit: "ratio", higherIsBetter: false },
      { key: "equity_to_assets", label: "Equity / assets", unit: "fraction", higherIsBetter: true },
      { key: "current_ratio", label: "Current ratio", unit: "ratio", higherIsBetter: true },
      { key: "quick_ratio", label: "Quick ratio", unit: "ratio", higherIsBetter: true },
      { key: "ebitda_interest", label: "Interest coverage", unit: "ratio", higherIsBetter: true },
      { key: "tangible_book_value_per_share", label: "Tangible book / share", unit: "usd", higherIsBetter: true },
    ],
  },
];

export const MARKET_FIELDS: FieldSpec[] = [
  { key: "price", label: "Price", unit: "usd" },
  { key: "market_cap", label: "Market cap", unit: "usdCompact" },
  { key: "pe_trailing", label: "P/E trailing", unit: "ratio" },
  { key: "pe_forward", label: "P/E forward", unit: "ratio" },
  { key: "peg_ratio", label: "PEG", unit: "ratio" },
  { key: "price_to_book", label: "P/B", unit: "ratio" },
  { key: "price_to_sales", label: "P/S", unit: "ratio" },
  { key: "price_to_tangible_book", label: "P/TBV", unit: "ratio" },
  { key: "ev_to_ebitda", label: "EV/EBITDA", unit: "ratio" },
  { key: "dividend_yield", label: "Dividend yield", unit: "fraction" },
  { key: "beta", label: "Beta", unit: "ratio" },
  { key: "week_52_high", label: "52w high", unit: "usd" },
  { key: "week_52_low", label: "52w low", unit: "usd" },
  { key: "avg_volume_30d", label: "Avg volume 30d", unit: "int" },
  { key: "analyst_target_price", label: "Analyst target", unit: "usd" },
];

/** Cary's nine, in the order the professor's Bloomberg code evaluates them. */
export const PIOTROSKI_LABELS: Record<string, string> = {
  roa_improved: "ROA improved",
  cfo_improved: "Operating cash flow improved",
  net_income_improved: "Net income improved",
  cfo_exceeds_net_income: "Cash flow exceeds net income",
  leverage_fell: "Leverage fell",
  current_ratio_improved: "Current ratio improved",
  shares_not_diluted: "No share dilution",
  gross_margin_improved: "Gross margin improved",
  asset_turnover_improved: "Asset turnover improved",
};

/** Fields we know this data plan cannot supply, so the page states it instead of showing a gap. */
export const KNOWN_ABSENT = [
  "Short interest",
  "Next-year EPS estimate",
  "Insider ownership",
  "Institutional ownership",
];
